### PR TITLE
Dashboards: Change parseUrlParam to take timezone into account

### PR DESCRIPTION
**Pull request by kaydelaney**
Fixes an issue where absolute time ranges didn't have their timezones taken into account, causing issues with time shift

### CODE REVIEW

**Review by joshhunt — COMMENTED** [review]
do you think this could be unit tested? 🥺

**Comment by juanicabanas** [comment]
Agree with @joshhunt . Is it possible to create some tests?

File: packages/grafana-data/src/datetime/datemath.test.ts
Comment (mckn) [inline packages/grafana-data/src/datetime/datemath.test.ts:61]: I'm not sure I agree that this code behaves correctly. But it can be my lack of understanding of what the `parse` function should do. The `Thursday, May 28, 2020 2:00:00 UTC` and `Thu May 28 2020 17:00:00 GMT+0300` is the same date. You are just displaying it in different time zones.

So there are no need to parse the date, in my opinion. If you want to display the date in the `Europe/Istanbul` you can just use the `dateToParse` and format in the timezone you would like to have.

If it would have been a string, that you wan to parse, without the time zone specified in it e.g. `Thursday, May 28, 2020 2:00:00` then this test would make sense because you don't know in what time zone the date is defined in. But when you create a date using the epoch it is defined in UTC.

Does this make sense?

**Review by mckn — COMMENTED** [review]
I'm not sure I agree with this code change. Can you elaborate a bit more on what you are trying to do and why it isn't working properly?

**Comment by kaydelaney** [comment]
> I'm not sure I agree with this code change. Can you elaborate a bit more on what you are trying to do and why it isn't working properly?

@mckn 
The issue arises when we have an absolute time range parsed from the URL, like `?from=1717127082395&to=1717148682395`. 

If we look at the `initTimeFromUrl` method on `TimeSrv`:
https://github.com/grafana/grafana/blob/287e3868ed251f1929f69daae3d7b47fd77b9c61/public/app/features/dashboard/services/TimeSrv.ts#L161-L167

https://github.com/grafana/grafana/blob/287e3868ed251f1929f69daae3d7b47fd77b9c61/public/app/features/dashboard/services/TimeSrv.ts#L125-L128

Notice that the `from` and `to` values are parsed as utc, and that the moment objects they're transformed to aren't aware of the time zone the dashboard is set to.  

The change in this PR stems from an issue where a panel with time shift "1d/d" isn't respected when the time range is absolute.
In the case of a time shift of "1d/d", the following code takes a "from" time, a "to" time, and rounds those times down and up respectively to the nearest day.
https://github.com/grafana/grafana/blob/287e3868ed251f1929f69daae3d7b47fd77b9c61/public/app/features/dashboard/utils/panel.ts#L147-L150

https://github.com/grafana/grafana/blob/287e3868ed251f1929f69daae3d7b47fd77b9c61/packages/grafana-data/src/datetime/datemath.ts#L172-L176

The crux of the problem is that the day boundaries are dependent on the time zone - information that isn't part of the `from` and `time` values.

By changing `parse` to take the passed timezone information into account, when we later perform the rounding, we correctly round to the day boundaries as they are in the user's timezone, rather than UTC.

**Comment by mckn** [comment]
I'm a bit unfamiliar with the `applyPanelTimeOverrides` functionality but I think I understand what the problem is.

A good practice when it comes to working with date time values is to always work with the value in UTC and only use time zones when displaying the date time value to the end user. Or when you are parsing the input from a user which probably enters the value in his/her current time zone. But between those places we should always work with date times in UTC.

When it comes to the time shift functionality in your example it doesn't matter if you are adding +1 day to a date that are in UTC or specified in a specific time zone as it will be shifted the same amount.  I guess where it can become a bit interesting is when you do a round up. Then you are not adding +1 day but instead adding +1 day and potentially a bit more depending on what time during the day it is in the target time zone.

So in this case I would probably have a function that calculate how much you should time shift depending on the time zone and then apply that time shift to the UTC date. But that is just how I would do it :)  With that said, I'm happy to approve this PR in its current state I'm just a bit worried that this change can have side effects in other places where you follow the "work with UTC" principle I tried to explain above.

It would be super to get input from the other reviewers of this PR. What are their thoughts?

**Review by mckn — APPROVED** [review]
I'm approving this PR but please see my previous comment and wait for at least one other approval.

File: packages/grafana-data/src/datetime/datemath.ts
Comment (axelavargas) [inline packages/grafana-data/src/datetime/datemath.ts:44]: if I understand correctly, before the `parse` function ignored the `timezone` argument, and returned a `moment` date with utc. 

>  * @param timezone Only string 'utc' is acceptable here, for anything else, local timezone is used.

It seems this was on purpose. If we change that behavior, it means that we can get side effects of people relying on expecting UTC, right? I wonder if we can tackle this issue differently 🤔

For instance, we could modify the `parseUrlParam` to handle `timezone`, instead of directly converting to UTC,  we could parse the epoch time and convert it to the specified timezone? 

```
 if (!isNaN(Number(value))) { 
   const epoch = parseInt(value, 10); 
    return moment.tz(epoch, timezone);
 } 

```

**Comment by diegommm** [comment]
Hi! Greetings from the Release Guild. Just letting you know I'm updating your PR's milestone to [11.2.x](https://github.com/grafana/grafana/milestone/529). If you need to see your changes in a previous version, please consider adding a backport label.

Thank you!

**Review by axelavargas — APPROVED** [review]
Awesome work @kaydelaney , the new approach works as expected 🥇. 

❤️  For adding the unit test! 

P.S. We might need to change the title of the PR , because we are not modifying anymore the `parse` function.

**Pull request merged**
